// @serenity-js/core/lib/events/ArtifactArchived.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ArtifactArchived = void 0;
const tiny_types_1 = require("tiny-types");
const io_1 = require("../io");
const model_1 = require("../model");
const screenplay_1 = require("../screenplay");
const DomainEvent_1 = require("./DomainEvent");
/**
 * @group Events
 */
class ArtifactArchived extends DomainEvent_1.DomainEvent {
    sceneId;
    name;
    type;
    path;
    artifactTimestamp;
    static fromJSON(o) {
        return new ArtifactArchived(model_1.CorrelationId.fromJSON(o.sceneId), model_1.Name.fromJSON(o.name), model_1.Artifact.ofType(o.type), io_1.Path.fromJSON(o.path), screenplay_1.Timestamp.fromJSON(o.artifactTimestamp), screenplay_1.Timestamp.fromJSON(o.timestamp));
    }
    constructor(sceneId, name, type, path, artifactTimestamp, timestamp) {
        super(timestamp);
        this.sceneId = sceneId;
        this.name = name;
        this.type = type;
        this.path = path;
        this.artifactTimestamp = artifactTimestamp;
        (0, tiny_types_1.ensure)('sceneId', sceneId, (0, tiny_types_1.isDefined)());
        (0, tiny_types_1.ensure)('name', name, (0, tiny_types_1.isDefined)());
        (0, tiny_types_1.ensure)('type', type, (0, tiny_types_1.isDefined)());
        (0, tiny_types_1.ensure)('path', path, (0, tiny_types_1.isDefined)());
        (0, tiny_types_1.ensure)('artifactTimestamp', artifactTimestamp, (0, tiny_types_1.isDefined)());
    }
    toJSON() {
        return {
            sceneId: this.sceneId.toJSON(),
            name: this.name.toJSON(),
            type: this.type.name,
            path: this.path.toJSON(),
            createdAt: this.artifactTimestamp.toJSON(),
            timestamp: this.timestamp.toJSON(),
            artifactTimestamp: this.artifactTimestamp.toJSON(),
        };
    }
}
exports.ArtifactArchived = ArtifactArchived;
